storage/innobase/include/trx0purge.h

```diff
@@ -55,80 +55,78 @@ Run a purge batch.
-
-	/** Compare two TrxUndoRsegs based on trx_no.
-	@param elem1 first element to compare
-	@param elem2 second element to compare
-	@return true if elem1 > elem2 else false.*/
-	bool operator()(const TrxUndoRsegs& lhs, const TrxUndoRsegs& rhs)
-	{
-		return(lhs.trx_no > rhs.trx_no);
-	}
-
-	/** Copy of trx_rseg_t::last_trx_no() */
-	trx_id_t trx_no= 0;
-private:
-	/** Rollback segments of a transaction, scheduled for purge. */
-	trx_rsegs_t m_rsegs{};
-};
-
-typedef std::priority_queue<
-	TrxUndoRsegs,
-	std::vector<TrxUndoRsegs, ut_allocator<TrxUndoRsegs> >,
-	TrxUndoRsegs>	purge_pq_t;
-
-/** Chooses the rollback segment with the oldest committed transaction */
-struct TrxUndoRsegsIterator {
-	/** Constructor */
-	TrxUndoRsegsIterator();
-	/** Sets the next rseg to purge in purge_sys.
-	Executed in the purge coordinator thread.
-	@retval false when nothing is to be purged
-	@retval true  when purge_sys.rseg->latch was locked */
-	inline bool set_next();
-
-private:
-	// Disable copying
-	TrxUndoRsegsIterator(const TrxUndoRsegsIterator&);
-	TrxUndoRsegsIterator& operator=(const TrxUndoRsegsIterator&);
-
-	/** The current element to process */
-	TrxUndoRsegs			m_rsegs;
-	/** Track the current element in m_rsegs */
-	TrxUndoRsegs::const_iterator	m_iter;
-};
-
 /** The control structure used in the purge operation */
 class purge_sys_t
 {
-  friend TrxUndoRsegsIterator;
+  /** Min-heap based priority queue over fixed size array */
+  class purge_queue_t
+  {
+    /** Array of indexes in trx_sys.rseg_array. */
+    alignas(CPU_LEVEL1_DCACHE_LINESIZE) byte m_array[TRX_SYS_N_RSEGS];
+    /** Pointer to the end of m_array. */
+    byte *m_end= m_array;
+
+  public:
+    struct trx_rseg_cmp
+    {
+      /** Compare two trx_rseg_t* based on trx_no.
+      @param lhs first index in trx_sys.rseg_array to compare
+      @param rhs second index in trx_sys.rseg_array to compare
+      @return whether lhs>rhs */
+      bool operator()(const byte lhs, const byte rhs)
+      {
+        ut_ad(lhs < TRX_SYS_N_RSEGS);
+        ut_ad(rhs < TRX_SYS_N_RSEGS);
+        /* We can compare without trx_rseg_t::latch, because rseg last
+        commit is always set before pushing rseg to purge queue. */
+        return trx_sys.rseg_array[lhs].last_trx_no() >
+               trx_sys.rseg_array[rhs].last_trx_no();
+      }
+    };
+    byte *begin() { return m_array; }
+    byte *end() { return m_end; }
+    const byte *c_begin() const { return m_array; }
+    const byte *c_end() const { return m_end; }
+    size_t size() const
+    {
+      size_t s= c_end() - c_begin();
+      ut_ad(s <= TRX_SYS_N_RSEGS);
+      return s;
+    }
+    bool empty() const { return !size(); }
+    void clear() { m_end= m_array; }
+
+    /** Push index of trx_sys.rseg_array into min-heap.
+    @param index index to push */
+    void push_index(byte index)
+    {
+      ut_ad(index < TRX_SYS_N_RSEGS);
+      ut_ad(size() + 1 <= TRX_SYS_N_RSEGS);
+      *(m_end++)= index;
```

Comment:
I think that this function needs to be `private`, or maybe merged with the public `push()`. The parentheses around the pointer post-increment are redundant.

The name `index` already has another meaning in the code base. I would suggest using just `i`, which we generally use for any array index.

Reply:
> I think that this function needs to be private

But it's used in `purge_sys_t::cleanse_purge_queue()`.